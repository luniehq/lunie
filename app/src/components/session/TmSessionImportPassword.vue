<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit.bind(this)">
      <h2 class="session-title">Choose password</h2>
      <div class="session-main bottom-indent">
        <Steps
          :steps="[`Recover`, `Name`, `Password`]"
          active-step="Password"
        />
        <TmFormGroup
          :error="$v.$error && $v.password.$invalid"
          field-id="import-password"
          field-label="Password"
        >
          <TmField
            id="import-password"
            v-model="password"
            type="password"
            placeholder="Must be at least 10 characters"
          />
          <TmFormMsg
            v-if="$v.password.$error && !$v.password.required"
            name="Password"
            type="required"
          />
          <TmFormMsg
            v-if="$v.password.$error && !$v.password.minLength"
            name="Password"
            type="minLength"
            min="10"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.passwordConfirm.$invalid"
          field-id="import-password-confirmation"
          field-label="Confirm Password"
        >
          <TmField
            id="import-password-confirmation"
            v-model="passwordConfirm"
            type="password"
            placeholder="Enter password again"
          />
          <TmFormMsg
            v-if="
              $v.passwordConfirm.$error && !$v.passwordConfirm.sameAsPassword
            "
            name="Password confirmation"
            type="match"
          />
          <TmFormMsg v-if="error" type="custom" :msg="errorMessage" />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Create" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required, minLength, sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import { mapState, mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"

export default {
  name: `session-import-password`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  data: () => ({
    error: false,
    errorMessage: ``,
  }),
  computed: {
    ...mapState([`recover`, `session`]),
    ...mapGetters([`network`, `networkSlug`, `isExtension`]),
    password: {
      get() {
        return this.$store.state.recover.password
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `password`, value })
      },
    },
    passwordConfirm: {
      get() {
        return this.$store.state.recover.passwordConfirm
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `passwordConfirm`, value })
      },
    },
  },
  beforeDestroy: function () {
    this.$store.dispatch(`resetRecoverData`)
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        await this.$store.dispatch(`createKey`, {
          seedPhrase: this.recover.seed,
          password: this.recover.password,
          name: this.recover.name,
          HDPath: this.session.HDPath,
          curve: this.session.curve,
          network: this.network,
        })
        if (this.isExtension) {
          this.$router.push(`/`)
        } else {
          this.$router.push({
            name: "portfolio",
            params: {
              networkId: this.networkSlug,
            },
          })
        }
      } catch (error) {
        this.error = true
        this.errorMessage = error.message
      }
    },
  },
  validations: () => ({
    password: { required, minLength: minLength(10) },
    passwordConfirm: { sameAsPassword: sameAs(`password`) },
  }),
}
</script>
