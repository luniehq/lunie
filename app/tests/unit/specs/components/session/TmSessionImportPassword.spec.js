import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import TmSessionImportPassword from "session/TmSessionImportPassword"
jest.mock(`scripts/google-analytics.js`, () => () => {})
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImportPassword`, () => {
  let wrapper, $store, getters

  beforeEach(() => {
    getters = {
      network: "lunie-net",
      networkSlug: "lunie",
      isExtension: false,
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false,
        },
        session: {
          accountType: ``,
          accountTypeIndex: ``,
        },
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(),
      mutations: {
        updateField: jest.fn(),
      },
    }
    wrapper = shallowMount(TmSessionImportPassword, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`validation should fail if passwords do not match`, async () => {
    wrapper.vm.$store.state.recover.password = `1234567890`
    wrapper.vm.$store.state.recover.passwordConfirm = `notthesame`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.passwordConfirm.$error).toBe(true)
    expect(wrapper.vm.$v.passwordConfirm.sameAsPassword).toBe(false)
  })

  it(`validation should not fail if passwords match`, async () => {
    wrapper.vm.$store.state.recover.password = `1234567890`
    wrapper.vm.$store.state.recover.passwordConfirm = `1234567890`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.passwordConfirm.$error).toBe(false)
    expect(wrapper.vm.$v.passwordConfirm.sameAsPassword).toBe(true)
  })

  it(`should commit updateField on password field change`, async () => {
    wrapper.setData({ password: `1234567890` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `password`,
      value: `1234567890`,
    })
  })

  it(`should commit updateField on passwordConfirm field change`, async () => {
    wrapper.setData({ passwordConfirm: `1234567890` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `passwordConfirm`,
      value: `1234567890`,
    })
  })

  it(`should dispatch createKey if passwords validate`, async () => {
    wrapper.vm.$store.state.recover.password = `1234567890`
    wrapper.vm.$store.state.recover.passwordConfirm = `1234567890`
    await wrapper.vm.onSubmit()
    expect($store.dispatch).toHaveBeenCalledWith(`createKey`, {
      name: ``,
      password: `1234567890`,
      network: "lunie-net",
      seedPhrase: ``,
      HDPath: undefined,
      curve: undefined,
    })
  })

  it(`should go to / when submit the form`, async () => {
    wrapper.vm.$store.state.recover.password = `1234567890`
    wrapper.vm.$store.state.recover.passwordConfirm = `1234567890`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      name: "portfolio",
      params: { networkId: "lunie" },
    })
  })

  it(`should show error on createKey dispatch error`, async () => {
    wrapper.vm.$store.state.recover.password = `1234567890`
    wrapper.vm.$store.state.recover.passwordConfirm = `1234567890`
    wrapper.vm.$store.dispatch = {
      createKey: jest.fn().mockRejectedValue(new Error()),
    }
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.error).toBe(true)
  })

  it(`should dispatch resetRecoverData when the component is destroyed`, async () => {
    wrapper.destroy()
    expect($store.dispatch).toHaveBeenCalledWith(`resetRecoverData`)
  })
})
