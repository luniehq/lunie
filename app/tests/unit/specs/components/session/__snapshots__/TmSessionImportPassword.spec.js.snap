// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImportPassword has the expected html structure 1`] = `
<sessionframe-stub
  icon=""
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Choose password
    </h2>
     
    <div
      class="session-main bottom-indent"
    >
      <steps-stub
        activestep="Password"
        steps="Recover,Name,Password"
      />
       
      <tmformgroup-stub
        fieldid="import-password"
        fieldlabel="Password"
      >
        <tmfield-stub
          id="import-password"
          placeholder="Must be at least 10 characters"
          type="password"
          value=""
        />
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-password-confirmation"
        fieldlabel="Confirm Password"
      >
        <tmfield-stub
          id="import-password-confirmation"
          placeholder="Enter password again"
          type="password"
          value=""
        />
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        customclass=""
        value="Create"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
